### PR TITLE
Components: remove lodash from `context/getStyledClassName`:

### PR DESCRIPTION
**Pull request by flootr**
## What?

This PR removes Lodash's `_.kebabCase()` from the `getStyledClassName` function used in ui/context. 

## Why?

Lodash is known to unnecessarily inflate the bundle size of packages, and in most cases, it can be replaced with native language functionality. See these for more information and rationale:

* https://github.com/WordPress/gutenberg/issues/16938#issuecomment-602837246
* https://github.com/WordPress/gutenberg/issues/17025
* https://github.com/WordPress/gutenberg/issues/39495 

## How?

Replaces `_.kebabCase` with `paramCase` from the `change-case` package (already used in the components package).

## Testing Instructions

1. Verify unit tests still pass; this change should be covered there
2. Start storybook locally and verify the story for the `ContextSystemProvider` works and looks as expected (compare with trunk here: [ContextSystemProvider](https://wordpress.github.io/gutenberg/?path=/story/components-experimental-contextsystemprovider--default))

### CODE REVIEW

File: packages/components/src/ui/context/get-styled-class-name-from-key.ts
Comment (tyxla) [inline packages/components/src/ui/context/get-styled-class-name-from-key.ts:4]: I may be wrong here, but I think some incorrect types might be deceiving us here.

See this usage of `getStyledClassNameFromKey()`:

https://github.com/WordPress/gutenberg/blob/8dee3e6764da100b20ad9abd41af5df086b5f42f/packages/components/src/ui/context/context-connect.ts#L94-L106

Seems like `namespace` could be an array? If that happens, this will crash, while `_.kebabCase( [ 'test', 'x' ] )` will work just fine, returning `test-x`.

This other use of `getStyledClassNameFromKey()` is in a similar situation like the one I [reported here](https://github.com/WordPress/gutenberg/pull/48637#issuecomment-1453564230):

https://github.com/WordPress/gutenberg/blob/cb10bcffde12b2b9196ae817285d3cbab684ddee/packages/components/src/ui/context/use-context-system.js#L54

Since the namespace is declared by externally exposed API (`useContextSystem()`), theoretically, the namespace could be `test1` and then one of them would yield  a `test1` class, and the other would yield a `test-1` class. That would basically be an unexpected breaking change.

I think to get around this, we need to make sure that the replacement function behaves exactly the same way when mixing letters, numbers and special characters.

Comment (flootr) [reply packages/components/src/ui/context/get-styled-class-name-from-key.ts:4]: Thanks Marin! I'm currently working through your feedback. Regarding

> Seems like namespace could be an array?

`namespace` is typed as `string` so if you try to pass an array to `contextConnect` you'd get a type error. I'm not exactly sure why the guard is still in place, it's probably a leftover from [#31099](https://github.com/WordPress/gutenberg/pull/31099) where the `Array<string>|string` type was replaced with `string`. My assumption is we could remove it.

Comment (tyxla) [reply packages/components/src/ui/context/get-styled-class-name-from-key.ts:4]: Sounds good for the array case. I'd suggest we remove it then, so we're intentional with our type approach here. Can be done in a separate PR, but I think it should be done before this one lands.

What do you think about this other difference that I mentioned? Reposting here for clarity:

> This other use of `getStyledClassNameFromKey()` is in a similar situation like the one I [reported here](https://github.com/WordPress/gutenberg/pull/48637#issuecomment-1453564230):
> 
> https://github.com/WordPress/gutenberg/blob/cb10bcffde12b2b9196ae817285d3cbab684ddee/packages/components/src/ui/context/use-context-system.js#L54
> 
> Since the namespace is declared by externally exposed API (`useContextSystem()`), theoretically, the namespace could be `test1` and then one of them would yield  a `test1` class, and the other would yield a `test-1` class. That would basically be an unexpected breaking change.
> 
> I think to get around this, we need to make sure that the replacement function behaves exactly the same way when mixing letters, numbers and special characters.

Comment (ciampo) [reply packages/components/src/ui/context/get-styled-class-name-from-key.ts:4]: > Since the namespace is declared by externally exposed API (useContextSystem())

I had a quick look and it doesn't seem that neither `getStyledClassNameFromKey` nor `useContextSystem()`  are exported by the `@wordpress/components` package (via the `packages/components/src/index.js` file).

Maybe @sarayourfriend could help us to clarify any remaining doubts on this, since she worked extensively on the context system?

Comment (tyxla) [reply packages/components/src/ui/context/get-styled-class-name-from-key.ts:4]: > I had a quick look and it doesn't seem that neither getStyledClassNameFromKey nor useContextSystem() are exported by the @wordpress/components package (via the packages/components/src/index.js file).

I've manually audited the existing usages and I have no further concern. The use case I've mentioned does not occur, and it won't cause any regressions, so we should be safe to move forward. 

The only suggestion I'd have before shipping this one would be removing the unnecessary array handling that we discussed above.

Comment (flootr) [reply packages/components/src/ui/context/get-styled-class-name-from-key.ts:4]: I started looking into removing the array handling we talked about and found a few more things which might be worth addressing. I'm going to discuss those with @ciampo this week (as I don't fully understand a few parts of the components context system). Could we unblock this PR and make your suggestion a follow up @tyxla?

Comment (tyxla) [reply packages/components/src/ui/context/get-styled-class-name-from-key.ts:4]: Cool, a follow-up sounds good 👍 Thanks @flootr

**Review by tyxla — APPROVED** [review]
LGTM 🚀

File: packages/components/src/ui/context/get-styled-class-name-from-key.ts
Comment (sarayourfriend) [reply packages/components/src/ui/context/get-styled-class-name-from-key.ts:4]: > Maybe @sarayourfriend could help us to clarify any remaining doubts on this, since she worked extensively on the context system?

Happy to help if I can, but can you clarify what the question is?

Comment (tyxla) [reply packages/components/src/ui/context/get-styled-class-name-from-key.ts:4]: It's mostly about the naming of contexts in `useContextSystem()` and specifically, treating numbers. It was about whether if we call `useContextSystem( 'test-1' )`, we should expect to get a className of `test-1` or `test1`. 

I realize this might be coming from the original g2 implementation, in which case @sarayourfriend might not have the answer too. 

I'm fine with making a decision here though, since `useContextSystem()` is still internal and uses no numbers in context names just yet. That's why I approved the PR.

Comment (sarayourfriend) [reply packages/components/src/ui/context/get-styled-class-name-from-key.ts:4]: > I realize this might be coming from the original g2 implementation, in which case @sarayourfriend might not have the answer too.

Yup, I have no idea what the original intention is.

> since useContextSystem() is still internal and uses no numbers in context names just yet

Excellent :grin:

**Review by ciampo — APPROVED** [review]
> I started looking into removing the array handling we talked about and found a few more things which might be worth addressing. I'm going to discuss those with @ciampo this week (as I don't fully understand a few parts of the components context system).

After discussing with Markus, we decided to keep the changes in this as lean as possible, in order to unblock further work on the lodash migration.

Markus may open a draft PR with some of his WIP about simplifying the Context System, but we won't realistically get back at it before completing the lodash and the TypeScript migration. After that, we will take a moment to assess the Context APIs, and decide the next steps (are these APIs really useful? Should we therefore use them more, or should we instead remove them entirely? If we keep them, what code optimisations can we apply?)

**Pull request merged**
